src/Symfony/Component/DependencyInjection/Loader/FileLoader.php

```diff
@@ -34,4 +36,72 @@ public function __construct(ContainerBuilder $container, FileLocatorInterface $l
 
         parent::__construct($locator);
     }
+
+    /**
+     * @experimental in version 3.3
+     */
+    public function registerClasses(Definition $prototype, $namespace, $directoryGlob)
+    {
+        if ('\\' !== substr($namespace, -1)) {
+            throw new InvalidArgumentException(sprintf('Namespace prefix must end with a "\\": %s.', $namespace));
+        }
+        if (!preg_match('/^(?:[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*+\\\\)++$/', $namespace)) {
+            throw new InvalidArgumentException(sprintf('Namespace is not a valid PSR-4 prefix: %s.', $namespace));
+        }
+
+        if (strlen($directoryGlob) === $i = strcspn($directoryGlob, '*?{[')) {
+            $directoryPrefix = $directoryGlob;
+            $directoryGlob = '';
+        } else {
+            $directoryPrefix = dirname(substr($directoryGlob, 0, 1 + $i));
+            $directoryGlob = substr($directoryGlob, strlen($directoryPrefix));
+        }
+
+        $directoryPrefix = $this->locator->locate($directoryPrefix, $this->currentDir, true);
+        $directoryPrefix = realpath($directoryPrefix) ?: $directoryPrefix;
+        $directoryGlob = $directoryPrefix.$directoryGlob;
+
+        // track directories only for new & removed files
+        $this->container->fileExists($directoryPrefix, '/^$/');
+
+        $classes = $this->findClasses($namespace, $directoryGlob, strlen($directoryPrefix));
+
+        foreach ($classes as $class) {
+            $this->container->setDefinition($class, clone $prototype);
```

Comment:
Does cloning a Definition behaves properly for other Definition objects (or other kind of objects) nested inside it (in arguments for instance).
If nested objects keep being shared, it would create weird things when compiler passes alter them later (as they will find the same instance for both and so changes will be shared). We need to ensure that we are deep-cloning here as necessary. I'm not sure the value objects of the component deal with it properly today